File: dev/plugins/base/SearchAndReplace/src/SearchResultsModel.cpp
#include "SearchResultsModel.h"
#include "SearchThread.h"

SearchResultsModel::SearchResultsModel( SearchThread* searchThread, QObject* parent )
	: QAbstractItemModel( parent )
{
	Q_ASSERT( searchThread );
	mRowCount = 0;
	mSearchThread = searchThread;

	// connections
	connect( mSearchThread, SIGNAL( reset() ), this, SLOT( thread_reset() ) );
	connect( mSearchThread, SIGNAL( resultsAvailable( const QString&, const SearchResultsModel::ResultList& ) ), this, SLOT( thread_resultsAvailable( const QString&, const SearchResultsModel::ResultList& ) ) );
}

int SearchResultsModel::columnCount( const QModelIndex& parent ) const
{
	Q_UNUSED( parent );
	return 1;
}

QVariant SearchResultsModel::data( const QModelIndex& index, int role ) const
{
	if ( !index.isValid() )
	{
		return QVariant();
	}

	SearchResultsModel::Result* result = this->result( index );

	Q_ASSERT( result );

	switch ( role )
	{
		case Qt::DisplayRole:
		{
			QString text;

			// index is a root parent
			if ( mParentsList.value( index.row() ) == result )
			{
				const int count = rowCount( index );
				text = mSearchDir.relativeFilePath( result->fileName );
				text.append( QString( " (%1)" ).arg( count ) );
			}
			// index is a root parent child
			else
			{
				text = tr( "Line: %1, Column: %2: %3" ).arg( result->position.y() +1 ).arg( result->position.x() ).arg( result->capture );
			}

			return text;
		}
		case Qt::ToolTipRole:
		{
			return data( index, Qt::DisplayRole );
		}
		case Qt::CheckStateRole:
		{
			if ( flags( index ) & Qt::ItemIsUserCheckable )
			{
				return result->checkState;
			}

			break;
		}
	}

	return QVariant();
}

QModelIndex SearchResultsModel::index( int row, int column, const QModelIndex& parent ) const
{
	if ( row >= rowCount( parent ) || column >= columnCount( parent ) )
	{
		return QModelIndex();
	}

	SearchResultsModel::Result* result = this->result( parent );

	// parent is a root parent
	if ( result && mParentsList.value( parent.row() ) == result )
	{
		result = mResults.at( parent.row() ).at( row );
		return createIndex( row, column, result );
	}

	Q_ASSERT( !parent.isValid() );

	// parent is invalid
	return createIndex( row, column, mParentsList[ row ] );
}

QModelIndex SearchResultsModel::parent( const QModelIndex& index ) const
{
	if ( !index.isValid() )
	{
		return QModelIndex();
	}

	SearchResultsModel::Result* result = this->result( index );

	// index is a root parent
	if ( result && mParentsList.value( index.row() ) == result )
	{
		return QModelIndex();
	}

	Q_ASSERT( index.isValid() );

	result = mParents[ result->fileName ];
	const int row = mParentsList.indexOf( result );

	// index is a root parent child
	return createIndex( row, index.column(), result );
}

int SearchResultsModel::rowCount( const QModelIndex& parent ) const
{
	// root parents
	if ( !parent.isValid() )
	{
		return mRowCount;
	}

	return parent.parent().isValid() ? 0 : mResults.at( parent.row() ).count();
}

Qt::ItemFlags SearchResultsModel::flags( const QModelIndex& index ) const
{
	Qt::ItemFlags flags = QAbstractItemModel::flags( index );
	SearchAndReplace::Properties* properties = mSearchThread->properties();

	if ( properties->mode & SearchAndReplace::ModeFlagReplace )
	{
		flags |= Qt::ItemIsUserCheckable;
	}

	SearchResultsModel::Result* result = this->result( index );

	if ( result )
	{
		if ( !result->enabled )
		{
			flags &= ~Qt::ItemIsEnabled;
			flags &= ~Qt::ItemIsSelectable;
		}
	}

	return flags;
}

bool SearchResultsModel::hasChildren( const QModelIndex& parent ) const
{
	// root parents
	if ( !parent.isValid() )
	{
		return mRowCount != 0;
	}

	return parent.parent().isValid() ? false : !mResults.at( parent.row() ).isEmpty();
}

bool SearchResultsModel::setData( const QModelIndex& index, const QVariant& value, int role )
{
	SearchResultsModel::Result* result = this->result( index );
	bool ok = false;

	switch ( role )
	{
		case Qt::CheckStateRole:
		{
			ok = true;
			break;
		}
		case SearchResultsModel::EnabledRole:
		{
			result->enabled = value.toBool();
			ok = true;
			break;
		}
	}

	if ( role != Qt::CheckStateRole )
	{
		if ( ok )
		{
			emit dataChanged( index, index );
		}

		return ok;
	}

	const Qt::CheckState state = Qt::CheckState( value.toInt() );
	const QModelIndex pIndex = index.parent();
	const bool isParent = !pIndex.isValid();
	SearchResultsModel::Result* pResult = this->result( pIndex );

	Q_ASSERT( result );

	if ( isParent )
	{
		const int pRow = mParentsList.indexOf( result );
		int checkedCount = 0;

		// update all children to same state as parent
		foreach ( SearchResultsModel::Result* r, mResults.at( pRow ) )
		{
			if ( r->enabled )
			{
				r->checkState = state;
				checkedCount++;
			}
		}

		const QModelIndex left = index.child( 0, 0 );
		const QModelIndex right = index.child( rowCount( index ) -1, columnCount( index ) -1 );
		// update root parent children
		emit dataChanged( left, right );

		if ( state == Qt::Unchecked )
		{
			checkedCount = 0;
		}

		if ( ( checkedCount == 0 && state == Qt::Checked ) || result->checkState == state )
		{
			ok = false;
		}

		if ( ok )
		{
			result->checkState = state;
		}
	}
	else
	{
		const int pRow = mParentsList.indexOf( pResult );
		int count = 0;
		int checkedCount = 0;

		foreach ( SearchResultsModel::Result* r, mResults.at( pRow ) )
		{
			count++;

			if ( r->checkState == Qt::Checked )
			{
				checkedCount++;
			}
		}

		if ( state == Qt::Checked )
		{
			checkedCount++;
		}
		else
		{
			checkedCount--;
		}

		result->checkState = state;

		// update parent
		if ( checkedCount == 0 )
		{
			pResult->checkState = Qt::Unchecked;
		}
		else if ( checkedCount == count )
		{
			pResult->checkState = Qt::Checked;
		}
		else
		{
			pResult->checkState = Qt::PartiallyChecked;
		}

		// update root parent index
		emit dataChanged( pIndex, pIndex );
	}

	// update clicked index
	emit dataChanged( index, index );

	return ok;
}

QModelIndex SearchResultsModel::index( SearchResultsModel::Result* result ) const
{
	const QModelIndex index;
	int row = mParentsList.indexOf( result );

	if ( row != -1 )
	{
		return createIndex( row, 0, result );
	}
	else if ( result )
	{
		SearchResultsModel::Result* pResult = mParents.value( result->fileName );

		if ( pResult )
		{
			row = mParentsList.indexOf( pResult );

			if ( row != -1 )
			{
				row = mResults.at( row ).indexOf( result );
				return createIndex( row, 0, result );
			}
		}
	}

	return index;
}

SearchResultsModel::Result* SearchResultsModel::result( const QModelIndex& index ) const
{
	return index.isValid() ? static_cast<SearchResultsModel::Result*>( index.internalPointer() ) : 0;
}

const QList<SearchResultsModel::ResultList>& SearchResultsModel::results() const
{
	return mResults;
}

void SearchResultsModel::clear()
{
	if ( mRowCount == 0 )
	{
		return;
	}
	
	beginRemoveRows( QModelIndex(), 0, mRowCount -1 );
	
	foreach ( const SearchResultsModel::ResultList& results, mResults )
	{
		qDeleteAll( results );
	}
	
	mResults.clear();
	qDeleteAll( mParents );
	mParents.clear();
	mParentsList.clear();
	mRowCount = 0;
	
	endRemoveRows();
}

void SearchResultsModel::thread_reset()
{
	clear();
}

void SearchResultsModel::thread_resultsAvailable( const QString& fileName, const SearchResultsModel::ResultList& results )
{
	if ( mRowCount == 0 )
	{
		emit firstResultsAvailable();
	}
	
	SearchResultsModel::Result* result = mParents[ fileName ];
	SearchAndReplace::Properties* properties = mSearchThread->properties();
	
	if ( mRowCount == 0 )
	{
		mSearchDir.setPath( properties->searchPath );
	}

	if ( !result )
	{
		result = new SearchResultsModel::Result( fileName );
		result->checkable = properties->mode & SearchAndReplace::ModeFlagReplace;
		result->checkState = result->checkable ? Qt::Checked : Qt::Unchecked;

		beginInsertRows( QModelIndex(), mRowCount, mRowCount );
		mParents[ fileName ] = result;
		mParentsList << result;
		mRowCount++;
		mResults << results;
		endInsertRows();
	}
	else
	{
		const int pRow = mParentsList.indexOf( result );
		const int count = mResults.at( pRow ).count();
		const QModelIndex index = createIndex( pRow, 0, result );

		beginInsertRows( index, count, count +results.count() -1 );
		mResults[ pRow ] << results;
		endInsertRows();

		emit dataChanged( index, index );
	}
}

void SearchResultsModel::thread_resultsHandled( const QString& fileName, const SearchResultsModel::ResultList& results )
{
	SearchResultsModel::Result* pResult = mParents.value( fileName );

	Q_ASSERT( pResult );

	const int pRow = mParentsList.indexOf( pResult );
	SearchResultsModel::ResultList& children = mResults[ pRow ];
	const QModelIndex pIndex = createIndex( pRow, 0, pResult );

	// remove root parent children
	foreach ( SearchResultsModel::Result* result, results )
	{
		const int index = children.indexOf( result );
		beginRemoveRows( pIndex, index, index );
		delete children.takeAt( index );
		endRemoveRows();
	}

	// remove root parent
	if ( children.isEmpty() )
	{
		beginRemoveRows( QModelIndex(), pRow, pRow );
		mResults.removeAt( pRow );
		mParentsList.removeAt( pRow );
		delete mParents.take( fileName );
		mRowCount--;
		endRemoveRows();
	}
	else
	{
		pResult->checkState = Qt::Unchecked;
		emit dataChanged( pIndex, pIndex );
	}
}

File: enki/plugins/searchreplace/searchresultsmodel.py
"""
searchresultsmodel --- Model for search results
===============================================
"""

from PyQt5.QtCore import pyqtSignal, QAbstractItemModel, QDir, QModelIndex, Qt
from PyQt5.QtWidgets import QApplication

from enki.lib.htmldelegate import htmlEscape


class Result:  # pylint: disable=R0902
    """One found by search thread item. Consists coordinates and capture. Used by SearchResultsModel
    """

    def __init__(self, fileName, wholeLine, line, column, match):  # pylint: disable=R0913
        self.fileName = fileName
        self.wholeLine = wholeLine
        self.line = line
        self.column = column
        self.match = match
        self.checkState = Qt.Checked

    def text(self):  # pylint: disable=W0613
        """Displayable text of search result. Shown as line in the search results dock
        """
        beforeMatch = self.wholeLine[:self.column].lstrip()
        afterMatch = self.wholeLine[self.column + len(self.match.group(0)):].rstrip()

        if QApplication.instance().palette().base().color().lightnessF() > 0.5:
            backgroundColor = 'yellow'
            foregroundColor = 'black'
        else:
            backgroundColor = 'maroon'
            foregroundColor = 'white'

        return '<html>' \
            'Line: %d, Column: %d: %s' \
            '<font style=\'background-color: %s; color: %s\'>%s</font>' \
            '%s' \
               '</html>' % \
            (self.line + 1,
             self.column,
             htmlEscape(beforeMatch),
             backgroundColor,
             foregroundColor,
             htmlEscape(self.match.group(0)),
             htmlEscape(afterMatch))

    def tooltip(self):
        """Tooltip of the search result"""
        return self.wholeLine.strip()

    def hasChildren(self):
        """Check if QAbstractItem has children"""
        return False


class FileResults:
    """Object stores all items, found in the file
    """

    def __init__(self, baseDir, fileName, results):
        self.baseDir = baseDir
        self.fileName = fileName
        self.results = results
        self.checkState = Qt.Checked

    def __str__(self):
        """Convertor to string. Used for debugging
        """
        return '%s (%d)' % (self.fileName, len(self.results))

    def updateCheckState(self):
        """Update own checked state after checked state of child result changed or
        child result removed
        """
        if all([res.checkState == Qt.Checked for res in self.results]):  # if all checked
            self.checkState = Qt.Checked
        elif any([res.checkState == Qt.Checked for res in self.results]):  # if any checked
            self.checkState = Qt.PartiallyChecked
        else:
            self.checkState = Qt.Unchecked

    def text(self):
        """Displayable text of the file results. Shown as line in the search results dock
        baseDir is base directory of current search operation
        """
        return '%s (%d)' % (QDir(self.baseDir).relativeFilePath(self.fileName), len(self.results))

    def tooltip(self):
        """Tooltip of the item in the results dock
        """
        return self.fileName

    def hasChildren(self):
        """Check if item has children
        """
        return 0 != len(self.results)


class SearchResultsModel(QAbstractItemModel):
    """AbstractItemodel used for display search results in 'Search in directory' and 'Replace in directory' mode
    """
    firstResultsAvailable = pyqtSignal()

    def __init__(self, parent):
        """Constructor of SearchResultsModel class
        """
        QAbstractItemModel.__init__(self, parent)
        self._replaceMode = False

        self.fileResults = []  # list of FileResults

    def setReplaceMode(self, enabled):
        """When replace mode is enabled, all items are checkState
        """
        self._replaceMode = enabled
        if self.fileResults:
            self.dataChanged.emit(self.index(0, 0, QModelIndex()),
                                  self.index(len(self.fileResults) - 1,
                                             len(self.fileResults[-1].results) - 1,
                                             QModelIndex()))

    def index(self, row, column, parent):
        """See QAbstractItemModel docs
        """
        if row >= self.rowCount(parent) or column > self.columnCount(parent):
            return QModelIndex()

        if parent.isValid():  # index for result
            result = parent.internalPointer().results[row]
            return self.createIndex(row, column, result)
        else:  # need index for fileRes
            return self.createIndex(row, column, self.fileResults[row])

    def parent(self, index):
        """See QAbstractItemModel docs
        """
        if not index.isValid():
            return QModelIndex()

        if not isinstance(index.internalPointer(), Result):  # it is an top level item
            return QModelIndex()

        result = index.internalPointer()
        for row, fileRes in enumerate(self.fileResults):
            if fileRes.fileName == result.fileName:
                return self.createIndex(row, 0, fileRes)
        else:
            assert(0)

    def hasChildren(self, item):
        """See QAbstractItemModel docs
        """
        # root parents
        if item.isValid():
            return item.internalPointer().hasChildren()
        else:
            return len(self.fileResults) != 0

    def columnCount(self, parent):  # pylint: disable=W0613
        """See QAbstractItemModel docs
        """
        return 1

    def rowCount(self, parent):
        """See QAbstractItemModel docs
        """
        if not parent.isValid():  # root elements
            return len(self.fileResults)
        elif isinstance(parent.internalPointer(), Result):  # result
            return 0
        elif isinstance(parent.internalPointer(), FileResults):  # file
            return len(parent.internalPointer().results)
        else:
            assert(0)

    def flags(self, index):
        """See QAbstractItemModel docs
        """
        flags = QAbstractItemModel.flags(self, index)

        if self._replaceMode:
            flags |= Qt.ItemIsUserCheckable

        return flags

    def data(self, index, role):
        """See QAbstractItemModel docs
        """
        if not index.isValid():
            return None

        # Common code for file and result
        result = index.internalPointer()
        if role == Qt.DisplayRole:
            return result.text()
        elif role == Qt.ToolTipRole:
            return result.tooltip()
        elif role == Qt.CheckStateRole:
            if self.flags(index) & Qt.ItemIsUserCheckable:
                return result.checkState

        return None

    def setData(self, index, value, role):
        """See QAbstractItemModel docs
        This method changes checked state of the item.
        If file unchecked - we need uncheck all items,
        if item unchecked...
        """
        if isinstance(index.internalPointer(), Result):  # it is a Result
            if role == Qt.CheckStateRole:
                # update own state
                index.internalPointer().checkState = value
                self.dataChanged.emit(index, index)  # own checked state changed
                # update parent state
                fileRes = index.parent().internalPointer()
                assert(isinstance(fileRes, FileResults))
                fileRes.updateCheckState()
                self.dataChanged.emit(index.parent(), index.parent())  # parent checked state might be changed
        elif isinstance(index.internalPointer(), FileResults):  # it is a FileResults
            if role == Qt.CheckStateRole:
                fileRes = index.internalPointer()
                fileRes.checkState = value
                for res in fileRes.results:
                    res.checkState = value
                firstChildIndex = self.index(0, 0, index)
                lastChildIndex = self.index(len(fileRes.results) - 1, 0, index)
                self.dataChanged.emit(firstChildIndex, lastChildIndex)
        else:
            assert(0)
        return True

    def setCheckStateForAll(self, state):
        """Check all items
        """
        for fileRes in self.fileResults:
            fileRes.checkState = state
            for match in fileRes.results:
                match.checkState = state
        self.dataChanged.emit(self.createIndex(0, 0, self.fileResults[0]),
                              self.createIndex(len(self.fileResults) - 1, 0, self.fileResults[-1]))

    def isFirstMatchChecked(self):
        """Check if first file in the search results is expanded
        """
        file = self.index(0, 0, QModelIndex())
        match = file.child(0, 0)
        return match.data(Qt.CheckStateRole) == Qt.Checked

    def clear(self):
        """Clear all results
        """
        self.beginRemoveRows(QModelIndex(), 0, len(self.fileResults) - 1)
        self.fileResults = []
        self.endRemoveRows()

    def appendResults(self, fileResultList):
        """Handler of signal from the search thread.
        New result is available, add it to the model
        """
        if not self.fileResults:  # appending first
            self.firstResultsAvailable.emit()
        self.beginInsertRows(QModelIndex(),
                             len(self.fileResults),
                             len(self.fileResults) + len(fileResultList) - 1)
        self.fileResults.extend(fileResultList)
        self.endInsertRows()

    def onResultsHandledByReplaceThread(self, fileName, results):
        """Replace thread has processed result, need to it from the model
        """
        for index, fileRes in enumerate(self.fileResults):  # try to find FileResults
            if fileRes.fileName == fileName:  # found
                fileResIndex = self.createIndex(index, 0, fileRes)
                if len(results) == len(fileRes.results):  # removing all
                    self.beginRemoveRows(QModelIndex(), index, index)
                    self.fileResults.pop(index)
                    self.endRemoveRows()
                else:
                    for res in results:
                        resIndex = fileRes.results.index(res)
                        self.beginRemoveRows(fileResIndex, resIndex, resIndex)
                        fileRes.results.pop(resIndex)
                        self.endRemoveRows()
                    if not fileRes.results:  # no results left
                        self.beginRemoveRows(QModelIndex(), index, index)
                        self.fileResults.pop(index)
                        self.endRemoveRows()
                    else:
                        fileRes.updateCheckState()
                return
        else:  # not found
            assert(0)

    def matchesCount(self):
        """Get count of matches, stored by the model
        """
        return sum([len(fileRes.results) for fileRes in self.fileResults])

    def empty(self):
        """Check if have some items
        """
        return len(self.fileResults) == 0
